torch_xla_py/xla_dist.py

```diff
@@ -402,3 +409,364 @@ def get_cluster(self):
+
+    export_cmd = []
+    for k in env_vars:
+      export_cmd.append(['export', '{}={}'.format(k, env_vars[k])])
+    for kv in self.env_vars:
+      export_cmd.append(['export', '{}'.format(kv)])
+    return export_cmd
+
+  def _prepare_scripts(self, cmd):
+    worker_script_map = {}
+    for i in range(len(self._cluster._client_workers)):
+      script_path = self.SCRIPT_PATH_TMPL.format(pid=os.getpid(), worker=i)
+
+      # ex. script = [['conda', 'activate', 'pytorch'], ['python', 'train.py']]
+      script = []
+      script.extend(self._env_vars_cmd(i))
+      # Setup environment for non-interactive non-login shell over ssh
+      script.append(['.', '/etc/profile'])
+      if self.docker_image:
+        script.append(self._docker_run_cmd(cmd))
+      else:
+        if self.conda_env:
+          script.append(['conda', 'activate', self.conda_env])
+        script.append(cmd)
+
+      # ex. script_body = 'conda activate pytorch; python train.py'
+      script_body = '; '.join([' '.join(command) for command in script])
+      os.makedirs(os.path.dirname(script_path), exist_ok=True)
+      with open(script_path, 'w') as f:
+        f.write(script_body)
+      subprocess.call(['chmod', '+x', script_path])
+      worker_script_map[self._cluster._client_workers[i]] = script_path
+
+    return worker_script_map
+
+  def _scp_scripts(self, script_map):
+
+    def _gcloud_scp(script_path, client_worker):
+      scp_cmd = self._build_scp_cmd(script_path, os.path.basename(script_path),
+                                    client_worker)
+      proc = subprocess.Popen(
+          scp_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+      self._stream_logs(proc, client_worker)
+
+    threads = []
+    for i, client_worker in enumerate(script_map):
+      if i == 0:
+        # ssh keygen single time
+        _gcloud_scp(script_map[client_worker], client_worker)
+        continue
+      thread = threading.Thread(
+          target=_gcloud_scp, args=(
+              script_map[client_worker],
+              client_worker,
+          ))
+      thread.daemon = True
+      thread.start()
+      threads.append(thread)
+
+    for thread in threads:
+      thread.join()
+
+  def _cleanup(self, script_map):
+
+    def _cleanup_worker(script_path, client_worker):
+      rm_script = ['rm', '~/{}'.format(os.path.basename(script_path))]
+      rm_script_cmd = self._build_ssh_cmd(rm_script, client_worker)
+      self._run_remote_cmd(rm_script_cmd, client_worker)
+      subprocess.call(['rm', script_path])
+      if self.docker_image:
+        rm_container = ['docker', 'rm', '-f', self.docker_container]
+        rm_container_cmd = self._build_ssh_cmd(rm_container, client_worker)
+        self._run_remote_cmd(rm_container_cmd, client_worker)
+      rm_proc = ['pkill', '-u', self.DEFAULT_USER_NAME]
+      rm_proc_cmd = self._build_ssh_cmd(rm_proc, client_worker)
+      self._run_remote_cmd(rm_proc_cmd, client_worker, log=False)
```

Comment:
We do do that a few times, but I feel that making that into another API is might be too specific?

We could either:
```
  def _build_and_run_ssh(remote_cmd, client_worker):
    ...

  def _build_and_run_scp(local_path, remote_path, client_worker):
    ...
```

or

```
  def _build_and_run_cmd(client_worker, remote_cmd=None, local_path=None, remote_path=None):
    # Build n run ssh cmd if remote_cmd is not None
    # Build n run scp cmd if local/remote_path are not None
    # Raise error otherwise
```

Not sure if either of them really buys us much compared to what we have now. What do you think?

Reply:
I'd only do this:
```Python
def _build_and_run_ssh(remote_cmd, client_worker):
   ...
```